File: imagecarroussel2/src/main/java/com/inlacou/imagecarroussel/PicturesFragmentPagerAdapter.java
package com.inlacou.imagecarroussel;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.PagerAdapter;
import android.view.View;

import java.util.ArrayList;

/**
 * Created by inlacou on 3/05/16.
 */
public class PicturesFragmentPagerAdapter extends android.support.v4.app.FragmentStatePagerAdapter {
	private static final String DEBUG_TAG = PicturesFragmentPagerAdapter.class.getName();
	private final Callbacks mCallbacks;
	private int pageCount;
	private final ArrayList<String> urls;

	/** Constructor of the class */
	public PicturesFragmentPagerAdapter(FragmentManager fm, ArrayList<String> urls, Callbacks callbacks) {
		super(fm);
		for (int i=urls.size()-1; i>=0; i--){
			if(urls.get(i)!=null && urls.get(i).isEmpty()) urls.remove(i);
		}
		this.urls = urls;
		pageCount = urls.size();
		mCallbacks = callbacks;
	}

	/** This method will be invoked when a page is requested to create */
	@Override
	public Fragment getItem(final int position) {
		PageImageFragment myFragment = new PageImageFragment();
		Bundle data = new Bundle();
		data.putInt("current_page", position);
		data.putInt("max_pages", pageCount);
		data.putString("url", urls.get(position));
		myFragment.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View view) {
				mCallbacks.onItemClick(position);
			}
		});
		myFragment.setArguments(data);
		return myFragment;
	}

	/** Returns the number of pages */
	@Override
	public int getCount() {
		return pageCount;
	}

	@Override
	public int getItemPosition(Object object) {
		return PagerAdapter.POSITION_NONE;
	}

	public interface Callbacks {
		void onItemClick(int position);
	}
}

File: imagecarroussel2/src/main/java/com/inlacou/imagecarroussel/PicturesFragmentPagerAdapter.kt
package com.inlacou.imagecarroussel

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.view.PagerAdapter

import java.util.ArrayList

/**
 * Created by inlacou on 3/05/16.
 */
class PicturesFragmentPagerAdapter @JvmOverloads
constructor(
		fragmentManager: FragmentManager,
		private val urls: ArrayList<String>,
		private val positionDisplay: PositionDisplayMode,
		private val showTopShadow: Boolean,
		private val infinite: Boolean,
		private val onClick: ((Int) -> Unit)? = null)
	: android.support.v4.app.FragmentStatePagerAdapter(fragmentManager) {

	private val pageCount: Int

	companion object {
		const val INITIAL_MAX_VALUE = 1000
	}

	var maxValue: Int = 0
		private set

	init {
		for (i in urls.indices.reversed()) {
			if (urls[i].isEmpty()) urls.removeAt(i)
		}
		pageCount = urls.size
		calculateMaxValue()
	}

	override fun notifyDataSetChanged() {
		calculateMaxValue()
		super.notifyDataSetChanged()
	}

	private fun calculateMaxValue(){
		if(pageCount>INITIAL_MAX_VALUE){
			pageCount
		}else{
			var maxVal = pageCount
			while (maxVal< INITIAL_MAX_VALUE){
				maxVal += pageCount
			}
			maxVal
		}
	}

	/** This method will be invoked when a page is requested to create  */
	override fun getItem(position: Int): Fragment {
		val virtualPos = if(infinite) position%pageCount else position
		val myFragment = PageImageFragment()
		val data = Bundle()
		data.putInt("current_page", virtualPos)
		data.putInt("max_pages", pageCount)
		data.putBoolean("showTopShadow", showTopShadow)
		data.putInt("positionDisplay", positionDisplay.ordinal)
		data.putString("url", urls[virtualPos])
		myFragment.onClickListener = { onClick?.invoke(virtualPos) }
		myFragment.arguments = data
		return myFragment
	}

	/** Returns the number of pages  */
	override fun getCount(): Int {
		return if(infinite) {
			if(pageCount==0) pageCount else maxValue
		}else {
			pageCount
		}
	}

	override fun getItemPosition(`object`: Any): Int {
		return PagerAdapter.POSITION_NONE
	}
}
